### PR TITLE
Search for configured soundfont and MT-32 ROMs in XDG_DATA_DIRS

### LINKED ISSUE #780: Consider using $XDG_DATA_HOME or  $HOME/.local/share for roms
**Is your feature request related to a problem? Please describe.**
No functional problem.  This is in response to the commit e73b18b46566327ad52039ba7620b93db9b8cb45

**Describe the solution you'd like**
The XDG Base Directory Specification has a solution for placement of data files.  Using the configuration location is not the correct place for anything but configuration files.
Dosbox should fully support the various directory specifications that are useful and available.
https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

**Describe alternatives you've considered**
Alternative is to ignore the specification and use anything else.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by kcgen**
Also includes searching `XDG_DATA_HOME/dosbox/{soundfont,mt32-roms}` subdirectories.

The commit messages include a real-world example of directories and paths opened in priority order.

Note: this PR doesn't create a `dosbox/*` directory tree under `XDG_DATA_HOME/` to reduce proliferation of empty directories.

Fixes #780 ; thank you @bmunger for catching this!

### CODE REVIEW

**Comment by kcgen** [comment]
This PR required adding a generic `split(..)` string-tokenizer function to process the `XDG_DATA_DIRS=` value.

I've added unit tests for; feel free to review and see if any other corner-cases should be tested.

``` text
[----------] 8 tests from Support_split
[ RUN      ] Support_split.NoBoundingDelims
[       OK ] Support_split.NoBoundingDelims (0 ms)
[ RUN      ] Support_split.DelimAtStartNotEnd
[       OK ] Support_split.DelimAtStartNotEnd (0 ms)
[ RUN      ] Support_split.DelimAtEndNotStart
[       OK ] Support_split.DelimAtEndNotStart (0 ms)
[ RUN      ] Support_split.DelimsAtBoth
[       OK ] Support_split.DelimsAtBoth (0 ms)
[ RUN      ] Support_split.MultiInternalDelims
[       OK ] Support_split.MultiInternalDelims (0 ms)
[ RUN      ] Support_split.MultiBoundingDelims
[       OK ] Support_split.MultiBoundingDelims (0 ms)
[ RUN      ] Support_split.MixedDelims
[       OK ] Support_split.MixedDelims (0 ms)
[ RUN      ] Support_split.Empty
[       OK ] Support_split.Empty (0 ms)
[----------] 8 tests from Support_split (0 ms total)
```

File: tests/support.cpp [hunk 1]
```diff
@@ -82,4 +82,64 @@ TEST(DriveIndex, DriveZ)
...
+
+TEST(Support_split, Empty)
+{
+	const std::vector<std::string> expected;
+	EXPECT_EQ(split("", ':'), expected);
+	EXPECT_EQ(split(":", ':'), expected);
+	EXPECT_EQ(split("::", ':'), expected);
+	EXPECT_EQ(split(" ", ' '), expected);
+	EXPECT_EQ(split("  ", ' '), expected);
+	EXPECT_EQ(split("   ", ' '), expected);
```
Comment (dreamer) [inline tests/support.cpp:142]: I think this behaviour is incorrect, e.g. `split("::", ':')` should result in `{"", "", ""}`.

I believe you wanted to emulate Python's interface on how to handle string splits (good idea!), but python does it a little bit differently:
```python
>>> ':::'.split(':')
['', '', '', '']
>>> '   '.split(' ')
['', '', '', '']
>>> '   '.split()
[]

```

Comment (kcgen) [reply tests/support.cpp:142]: If a string only contains delimiters, then there is no content when they're removed.

On the Python side, I suspect that returning a <None> would throw wrenches into nice Pythonic loops :-) Then again, they could return an empty vector and avoid the None. 

Also, pedantically, if a delimiter separates content on both its sides, then if:

"content_1:content_2" produces `{"content_1", "content_2"}`

And if the absence of delimited content still has value, then therefore:

 - ":content_2" should produce {"", "content_2"}
 - "content_1:" should produce {"content_1", ""}
 - ":" should produce {"", ""}
 - ":::" should produce 5 empties

The point being that if there's value in reporting non-content on the left-and-right sides of a delimiter, then that should be true even in a minimal case.

But more pragmatically -- the only value I see here is possibly counting the invalid or wasted delimeters that were used.  The split() implementation would get less elegant and the user-side would get uglier checking for empty() and discarding them too.

Maybe if/when we have a need to understand non-content we could add this?

Comment (kcgen) [reply tests/support.cpp:142]: Actually, maybe we can implement it with a toggle that says "I care about non-content" versus "ignore non-content".

That said.. given the two scenarios - I honestly feel we only will ever care about the positive content and not the absence of content.

Comment (dreamer) [reply tests/support.cpp:142]: Oh, no - the implementation will be *easier* than it is now. When looking for next string to extract start with next character instead of looking for next non-delimiter. One-line change.

I suggest adding another version of this function, one that does *not* accept delimiter as parameter (assuming delimiter is any whitespace), and acts as the current implementation (matching Python's `str.split()` - without parameters).

Comment (kcgen) [reply tests/support.cpp:142]: Thanks @dreamer ; will do!

**Comment by kcgen** [comment]
Updates and new test cases inbound. 
 - `split(string, delim)` now returns absent content on either side of the delimiter as an empty string
 - Updated split on delimiter is here: https://github.com/dosbox-staging/dosbox-staging/pull/783/commits/bfadd2ce5ab8c1017c874f07da3bf62db4ce4057)
 - `split(string)` returns non-whitespace words, without absent content stripped
 - New split on whitespace is here: https://github.com/dosbox-staging/dosbox-staging/pull/783/commits/bfadd2ce5ab8c1017c874f07da3bf62db4ce4057

The MT-32 and FluidSynth `XDG_DATA_DIRS` still uses the first function, splitting on the `:`, however now skips past empty entries.

File: include/cross.h [hunk 2]
```diff
@@ -30,9 +30,25 @@
...
+#ifdef _CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES
+#undef _CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES
+#define _CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES 1
+#endif
+
+#ifdef _CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES_COUNT
+#undef _CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES_COUNT
+#define _CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES_COUNT 1
+#endif
+#else // _MSC_VER < 1600
```
Comment (dreamer) [inline include/cross.h:45]: We don't want to allow MSVC to automatically wrap the POSIX API. That's because safe wrappers are going to change the behaviour of the function and we *will* have regressions.

Even Microsof's documentation warns about this:

> Before doing so, however, make sure that your code passes the count of characters, not the size of the buffer (a common mistake). 

This is *exactly the opposite* of how most of code in DOSBox passes the count parameter here - and it is not a "common mistake" - it's by design. In fact every place where the code did *not* pass the buffer size is a potential buffer overflow issue (and we haven't finished the transition to remove all remaining cases IIRC).

Remove these defines and simply raise the limit of warnings for MSVC, please.

File: include/support.h [hunk 3]
```diff
@@ -185,6 +186,30 @@ void upcase(std::string &str);
 void lowcase(std::string &str);
 void strip_punctuation(std::string &str);
 
+// Split a string on an arbitrary character delimiter. Absent string content on
+// either side of a delimiter is treated as an empty string. For example:
+//   split("abc:", ':') returns {"abc", ""}
+//   split(":def", ':') returns {"", "def"}
+//   split(":", ':') returns {"", ""}
+//   split("::", ':') returns {"", "", ""}
+std::vector<std::string> split(std::string const &seq, const char delim);
```
Comment (dreamer) [inline include/support.h:195]:
```C++
std::vector<std::string> split(const std::string &seq, char delim);
```
Put const before type for reference and no const for parameters passed by-value in function declarations. Function definition can stay as it is - it's good :)

File: src/midi/midi_fluidsynth.cpp [hunk 4]
```diff
@@ -128,19 +128,35 @@ static std::vector<std::string> get_data_dirs()
...
-	return {
-	        CROSS_GetPlatformConfigDir() + "soundfonts/",
-	        xdg_data_home + "/soundfonts/",
-	        xdg_data_home + "/sounds/sf2/",
-	        "/usr/local/share/soundfonts/",
-	        "/usr/local/share/sounds/sf2/",
-	        "/usr/share/soundfonts/",
-	        "/usr/share/sounds/sf2/",
-	};
+	std::vector<std::string> dirs = {xdg_data_home + "/dosbox/soundfonts/",
```
Comment (dreamer) [inline src/midi/midi_fluidsynth.cpp:136]: The container is no longer being filled once during initialization… Which means that this way of std::vector usage is not appropriate any more. Two ways to move through this:

1. replace std::vector with std::deque, or:
2. reserve enough space in vector (it is possible to pre-calculate how many strings will end up in it in this case)

File: src/misc/support.cpp [hunk 5]
```diff
@@ -124,6 +124,36 @@ void trim(std::string &str)
 	str.erase(0, empty_pfx);
 }
 
+std::vector<std::string> split(const std::string &seq, const char delim)
```
Comment (dreamer) [inline src/misc/support.cpp:127]: Both versions of this function behave better now - thanks :)

Only now I recognized a problem with the version that's not using delim char - we cannot easily determine vector size to reserve before filling it. The ways to *not* solve this:

1. Do NOT use std::list

Three possible ways to address this:

1. We could use std::forward_list - not the perfect solution, but OK, and lower overhead than std::list
2. We could use std::deque - faster than std::list, but with high overhead for cases with small number of elements
3. Calculate the number of elements to reserve for std::vector. For function using delimiter it's simple - just count the number of delimiters in input string (single pass through the string is less work than several re-initializations of std::vector's backing buffer). For function splitting by space it's more complicated - we could calculate number of adjacent space/non-space characters… but maybe it's an overengineering at this point of implementation - it will be much easier to initially assume that the majority of `split` calls will generate small outputs and reserve something reasonable - e.g. start with reserving 5 elements.

**Review by dreamer — CHANGES_REQUESTED** [review]
Looks OK already, but still needs a little bit of improvement. Especially we can't use those MSVC defines…

File: include/support.h (hunk 3 above)
Comment (kcgen) [reply include/support.h:195]: Argh - I caught that yesterday, but maybe only updated in the .cpp and didn't get the header. Will do!

File: src/misc/support.cpp (hunk 5 above)
Comment (kcgen) [reply src/misc/support.cpp:127]: Thanks for this feedback @dreamer. I also like the count and reserve approach, given these functions might wind up being used at runtime too where performance becomes more important. Will take a crack at the run-length counts on the white space splitter; it shouldn't be too onerous.

File: include/cross.h (hunk 2 above)
Comment (kcgen) [reply include/cross.h:45]: Ugh.. what a mess Microsoft has made. Certainly agree that we don't want behavioral changes - so thanks for catching this. I'm not even sure why Microsoft would offer these then, if they're prone to unintended side effects.

Is there a rote way we can move off these en-mass, or can describe a couple simply rules (if using printf then handle the char* array like this, otherwise use std::string for strcat, and so on?). 

It would be nice to chip them away.. say 10 fixes a day over the course of a month or so.  Churns my stomach trying to wade through such a mess of legit warnings.

Comment (dreamer) [reply include/cross.h:45]: I have work-in-progress version of new set of `safe_*` functions… somewhere on an old branch (that's why `string_utils.h` got introduced).

I never finished the work because I wanted to have cross-platform unit-tests work first (specifically unit tests working on Windows).

File: src/misc/support.cpp (hunk 5 above)
Comment (dreamer) [reply src/misc/support.cpp:127]: Without actually running through the compiler - adjacent chars counter could look something like this:
```C++
size_t n = 1;
while (std::adjacent_find(begin(str), end(str), [](char a, char b) { return !isspace(a) && isspace(b); } != end(str))
        n += 1;
```

File: include/cross.h (hunk 2 above)
Comment (kcgen) [reply include/cross.h:45]: Right on.  If the `safe_*` versions include Windows-specific calls, then makes sense to back-burner it until we can thoroughly test it (will meson will unlock that for us? Or is it's a google test suite vcpkg thing?)

Otherwise if the code is pure C++11/14, maybe we can introduce it sooner (assuming we write similar unit tests and everything is passing on Linux and macOS)?

File: src/midi/midi_fluidsynth.cpp (hunk 4 above)
Comment (kcgen) [reply src/midi/midi_fluidsynth.cpp:136]: Went with deque for the directory lists.  Felt that keeping the code tighter (and avoiding pre-tally  code) is better for maintainers given there's zero performance concern here.

File: src/misc/support.cpp (hunk 5 above)
Comment (kcgen) [reply src/misc/support.cpp:127]: That one repeats the same search (so spins).  

This handles all the corner cases, like `" a b"`, `"a b "`, and `" a b "`.  The assignment of `head = ` and comparison to `seq.end()` can all be done inside the `while`, but it's borderline obfuscated at that point.

``` c++
    auto head = seq.begin();
    size_t n = !isspace(seq.back());
    while (head != seq.end()) {
        head = std::adjacent_find(head,  seq.end(), [](char a, char b) { return !isspace(a) && isspace(b); });
        if (head != seq.end()) {
            ++n;
            ++head;
        }
    }
```

Comment (kcgen) [reply src/misc/support.cpp:127]: I just went with the original pass through, but turned it into a count:
``` c++
	size_t n = 0;
	auto head = seq.find_first_not_of(whitespace, 0);
	while (head != std::string::npos) {
		const auto tail = seq.find_first_of(whitespace, head);
		head = seq.find_first_not_of(whitespace, tail);
		++n;
	}
```

Comment (kcgen) [reply src/misc/support.cpp:127]: Latest update asserts that the calculated number of reserved elements matches the actual elements populating the vector.

**Comment by kcgen** [comment]
Latest update tightens up the rom directory handling:
 - No longer passes the user's romdir into the `get_rom_dirs(..)` function and instead simply emplaces it to the front of the deque.
 - Calls `cross_resolve_home` on the user's romdir.

File: src/midi/midi_fluidsynth.cpp [hunk 6]
```diff
@@ -126,21 +127,39 @@ static std::vector<std::string> get_data_dirs()
 
 #else
 
-static std::vector<std::string> get_data_dirs()
+static std::deque<std::string> get_data_dirs()
```
Comment (dreamer) [inline src/midi/midi_fluidsynth.cpp:130]: Great choice :)

File: src/midi/midi_mt32.cpp [hunk 7]
```diff
@@ -237,21 +252,28 @@ bool MidiHandler_mt32::Open(MAYBE_UNUSED const char *conf)
...
-	// Load the ROMs for the selected model
+	// Get the user's configured ROM directory; otherwise use 'mt32-roms'
+	std::string preferred_dir = section->Get_string("romdir");
+	if (preferred_dir.empty()) // already trimmed
+		preferred_dir = "mt32-roms";
+	if (preferred_dir.back() != '/' && preferred_dir.back() != '\\')
+		preferred_dir += CROSS_FILESPLIT;
+
+	// Make sure we search the user's configured directory first
+	rom_dirs.emplace_front(CROSS_ResolveHome((preferred_dir)));
```
Comment (dreamer) [inline src/midi/midi_mt32.cpp:269]: Nice :)

File: tests/support.cpp [hunk 8]
```diff
@@ -82,4 +82,121 @@ TEST(DriveIndex, DriveZ)
 	EXPECT_EQ(25, drive_index('Z'));
 }
 
+TEST(Support_split_delim, NoBoundingDelims)
```
Comment (dreamer) [inline tests/support.cpp:85]: I though I made this comment earlier but I can't find it now…

An edge-case test is missing: splitting a non-empty string that has no characters for splitting, e.g.

`split("abc", 'x') -> {"abc"}`, `split("abc") -> {"abc"}`

**Review by dreamer — CHANGES_REQUESTED** [review]
Looks really good, and split implementation is top-notch now :)

I think this will be the last round of review.

File: tests/support.cpp (hunk 8 above)
Comment (kcgen) [reply tests/support.cpp:85]: (Good news is that the your prior comment showed up on my side and wasn't lost)

**Comment by kcgen** [comment]
> Looks really good, and split implementation is top-notch now :)
> 
> I think this will be the last round of review.

Thanks for the finishing touches, @dreamer.
Tests are passing and the synth-music is playing :-)

**Review by dreamer — APPROVED** [review]
@kcgen Thanks! :)
